rfcs/0008-storage-ref.md

```diff
@@ -0,0 +1,249 @@
+- Feature Name: Storage References
+- Start Date: 2023-02-10)
+- RFC PR: [FuelLabs/sway-rfcs#0008](https://github.com/FuelLabs/sway-rfcs/pull/023)
+- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/3043)
+
+# Summary
+
+[summary]: #summary
+
+One paragraph explanation of the feature.
+
+# Motivation
+
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected outcome?
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A storage reference stores a `b256` key that describes a storage slot:
+
+```rust
+pub struct StorageRef<T> {
+    key: b256,
+}
+```
+
+We can implement `store` and `get` as follows:
+
+```rust
+impl<T> StorageRef<T> {
+    #[storage(read)]
+    pub fn get(self) -> Option<T> {
```

Comment:
I can imagine getting frustrated working around this `Option` for references that you *know* must be valid, e.g. `storage.foo` in `storage { foo: u32 = 42, }`. Was the intent for this to be `pub fn get(self) -> T`?

I imagine there will be cases where you *would* want to return an `Option`, e.g. if you ever have a reference to an element of a storage *collection*, as it's possible the element may get removed between when you construct the reference and when `get` is called as we have no borrowcheck in Sway. Maybe there needs to be different types for the two cases? I.e. 1 where you know the element will always be valid and 2. another where the element may have been freed? :thinking:

Reply:
This is pretty annoying yes - when I started thinking about this, `get` did not return an option but now it does. What do you mean by two different types? Two different `StorageRef` types?

Reply:
I added this as a drawback and we should discuss how to make this more ergonomic.